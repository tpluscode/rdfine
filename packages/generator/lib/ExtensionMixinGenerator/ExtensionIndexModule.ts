import { GeneratedModule } from '../index'
import { Project } from 'ts-morph'

export class ExtensionIndexModule implements GeneratedModule {
  constructor(public prefix: string, public terms: string[]) {
  }

  writeModule({ project }: { project: Project }): void {
    const extensionsModule = project.createSourceFile(`extensions/${this.prefix}.ts`, {}, { overwrite: true })

    for (const extendedTerm of this.terms) {
      extensionsModule.addExportDeclaration({
        namedExports: [`${extendedTerm}MixinEx`],
        moduleSpecifier: `./${this.prefix}/${extendedTerm}`,
      })
    }
  }
}
