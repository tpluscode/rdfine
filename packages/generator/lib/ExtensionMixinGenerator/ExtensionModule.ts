import { NamespaceDeclarationKind, Project, SourceFile } from 'ts-morph'
import { GraphPointer } from 'clownface'
import { shrink } from '@zazuko/rdf-vocabularies'
import { ExternalResourceType, TypeMetaCollection } from '../types'
import { Context } from '../index'
import { JavascriptProperty } from '../property/JsProperties'
import { PropertyWriter } from '../property/PropertyWriter'
import { MixinModuleBase } from '../MixinGenerator/MixinModuleBase'

export class ExtensionModule extends MixinModuleBase<ExternalResourceType> {
  properties: JavascriptProperty[];
  private extended: { prefix: string; term: string }
  private interfaceName: string

  constructor({ domain, type }: { domain: GraphPointer; type: ExternalResourceType }) {
    super(domain, type)
    this.properties = []
    const [prefix, term] = shrink(this.node.value).split(':')
    this.extended = { prefix, term }
    this.interfaceName = `${this.extended.term}Ex`
  }

  addProperties(...properties: JavascriptProperty[]) {
    for (const property of properties) {
      this.properties.push(property)
    }
  }

  writeModule({ project, context }: { project: Project; types: TypeMetaCollection; context: Omit<Context, 'properties'>; indexModule: SourceFile }): void {
    const module = project.createSourceFile(`extensions/${this.extended.prefix}/${this.extended.term}.ts`, {}, { overwrite: true })

    context.log.debug(`Generating extension mixin for ${this.type.qualifiedName}`)

    const interfaceEx = module.addInterface({
      name: `${this.interfaceName}<D extends RDF.DatasetCore = RDF.DatasetCore>`,
    })

    this.augmentExtendedModule(module)

    const { mixinClass, mixinFunction } = this.createMixinFunction(module, context)

    const propertyWriter = new PropertyWriter({
      interfaceDeclaration: interfaceEx,
      classDeclaration: mixinClass,
      context,
      module: this,
    })
    this.properties.forEach(propertyWriter.addProperty.bind(propertyWriter))

    const nsBuilderTerm = this.extended.term === this.type.localName
      ? `${this.extended.prefix}.${this.extended.term}`
      : `${this.extended.prefix}['${this.extended.term}']`
    module.addStatements([
      `${mixinFunction.getName()}.appliesTo = ${nsBuilderTerm}`,
    ])

    this.addRdfineNamespaceImport(this.type)
    this.addImports(module, context)
  }

  private augmentExtendedModule(module: SourceFile) {
    const augmentedModule = module.addNamespace({
      name: `'${this.type.module}'`,
    }).setDeclarationKind(NamespaceDeclarationKind.Module)

    augmentedModule.addInterface({
      name: `${this.type.localName}<D extends RDF.DatasetCore>`,
      extends: [`${this.interfaceName}<D>`],
    })
  }

  private createMixinFunction(mixinFile: SourceFile, context: Omit<Context, 'properties'>) {
    const mixinFunction = mixinFile.addFunction({
      name: `${this.type.localName}MixinEx`,
      typeParameters: [{
        name: 'Base',
        constraint: `ExtendingConstructor<${this.type.qualifiedName}, ${this.interfaceName}>`,
      }],
      parameters: [{
        name: 'Resource',
        type: 'Base',
      }],
      returnType: `Constructor<${this.interfaceName} & RdfResourceCore> & Base`,
      isExported: true,
    })

    const mixinClass = mixinFunction.addClass({
      name: 'Impl',
      extends: 'Resource',
      implements: [this.interfaceName],
    })

    mixinClass.addDecorator({
      name: 'namespace',
      arguments: [context.prefix],
    })

    mixinFunction.addStatements('return Impl')

    return { mixinClass, mixinFunction }
  }

  private addImports(mixinFile: SourceFile, context: Omit<Context, 'properties'>) {
    mixinFile.addImportDeclaration({
      namedImports: ['ExtendingConstructor', 'Constructor', 'namespace', 'property'],
      moduleSpecifier: '@tpluscode/rdfine',
    })
    mixinFile.addImportDeclaration({
      namedImports: ['RdfResourceCore'],
      moduleSpecifier: '@tpluscode/rdfine/RdfResource',
      isTypeOnly: true,
    })

    mixinFile.addImportDeclaration({
      namespaceImport: '$rdf',
      moduleSpecifier: '@rdf-esm/data-model',
    })
    mixinFile.addImportDeclaration({
      namespaceImport: 'RDF',
      moduleSpecifier: 'rdf-js',
      isTypeOnly: true,
    })
    mixinFile.addImportDeclaration({
      namedImports: [this.extended.prefix],
      moduleSpecifier: '@tpluscode/rdf-ns-builders',
    })
    mixinFile.addImportDeclaration({
      namedImports: [context.prefix],
      moduleSpecifier: '../../lib/namespace',
    })
    mixinFile.addImportDeclaration({
      namespaceImport: context.defaultExport,
      moduleSpecifier: '../..',
      isTypeOnly: true,
    })

    Object.entries(this.namespaceImports)
      .forEach(([moduleSpecifier, namespaceImport]) => {
        mixinFile.addImportDeclaration({
          moduleSpecifier,
          namespaceImport,
          isTypeOnly: true,
        })
      })

    this.mixinImports
      .filter(imported => imported.type === 'ExternalResource')
      .forEach(imported => {
        if (imported.type === 'Resource') {
          mixinFile.addImportDeclaration({
            moduleSpecifier: './' + imported.localName,
            namedImports: [imported.mixinName],
          })
        } else {
          const superImport = mixinFile.addImportDeclaration({
            moduleSpecifier: imported.module,
          })

          superImport.addNamedImport({
            name: imported.exportName,
            alias: imported.mixinName,
          })
        }
      })
  }
}
