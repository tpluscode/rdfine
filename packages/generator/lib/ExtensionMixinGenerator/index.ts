import { rdfs } from '@tpluscode/rdf-ns-builders'
import { TypeMetaCollection } from '../types'
import { Context, GeneratedModule } from '../index'
import { prefixes, shrink } from '@zazuko/rdf-vocabularies'
import TermMap from '@rdfjs/term-map'
import { NamedNode } from 'rdf-js'
import { GraphPointer } from 'clownface'
import { ExtensionModule } from './ExtensionModule'
import { toJavascriptProperties } from '../property/JsProperties'
import { findRanges } from '../property'
import { ExtensionIndexModule } from './ExtensionIndexModule'

interface AddToMapParams {
  property: GraphPointer
  domain: GraphPointer
  types: TypeMetaCollection
  context: Pick<Context, 'properties' | 'log'>
}

function addExtension(extensions: Map<NamedNode, ExtensionModule>, { property, domain, types, context }: AddToMapParams) {
  const type = types.get(domain)

  if (!(domain.term.termType === 'NamedNode' && type?.type === 'ExternalResource')) {
    return false
  }

  let module = extensions.get(domain.term)
  if (!module) {
    module = new ExtensionModule({ domain, type })
    extensions.set(domain.term, module)
  }
  module.addProperties(...toJavascriptProperties(property, findRanges(property), types, context))

  return true
}

export function findTermsToGenerate(types: TypeMetaCollection, context: Pick<Context, 'vocabulary' | 'log' | 'properties' | 'prefix'>): GeneratedModule[] {
  const extensions = new TermMap<NamedNode, ExtensionModule>()
  const extendedTerms = new Map<string, Set<string>>()

  const extendingProperties = context.vocabulary.has(rdfs.domain)
  for (const property of extendingProperties.toArray()) {
    for (const domain of property.out(rdfs.domain).toArray()) {
      const [prefix, term] = shrink(domain.value).split(':')
      if (domain.value.startsWith(prefixes[context.prefix])) {
        continue
      }

      if (!types.get(domain)) {
        context.log.debug(`Skipping extensions of ignored type ${domain.value}`)
        continue
      }

      if (!(prefix in prefixes)) {
        context.log.warn(`Extended term from unknown vocabulary: ${domain.value}`)
        continue
      }

      const added = addExtension(extensions, {
        property,
        domain,
        types,
        context,
      })

      if (added) {
        const extendedTermFromPrefix = extendedTerms.get(prefix) || new Set()
        extendedTermFromPrefix.add(term)
        extendedTerms.set(prefix, extendedTermFromPrefix)
      }
    }
  }

  const indexModules = [...extendedTerms].map(([prefix, terms]) => new ExtensionIndexModule(prefix, [...terms]))

  return [
    ...extensions.values(),
    ...indexModules,
  ]
}
