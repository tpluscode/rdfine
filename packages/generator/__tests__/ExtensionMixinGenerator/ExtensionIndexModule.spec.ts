import { Project } from 'ts-morph'
import { ExtensionIndexModule } from '../../lib/ExtensionMixinGenerator/ExtensionIndexModule'

describe('ExtensionIndexModule', () => {
  it('generates exports for extended terms', () => {
    // given
    const project = new Project({
      useInMemoryFileSystem: true,
    })
    const module = new ExtensionIndexModule('rdfs', ['Class', 'Property'])

    // when
    module.writeModule({ project })

    // then
    expect(project.getSourceFile('extensions/rdfs.ts')).toMatchSnapshot()
  })
})
