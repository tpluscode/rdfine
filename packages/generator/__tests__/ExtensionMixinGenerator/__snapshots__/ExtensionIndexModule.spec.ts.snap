// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExtensionIndexModule generates exports for extended terms: toMatchSnapshot 1`] = `
"export { ClassMixinEx } from \\"./rdfs/Class\\";
export { PropertyMixinEx } from \\"./rdfs/Property\\";
"
`;
